drivers/lkss/lab2/ex5/gpio_led_blink.c

```diff
@@ -0,0 +1,170 @@
+{
+	char kbuf[8];
+
+	if (count >= sizeof(kbuf))
+		return -EINVAL;
+
+	if (copy_from_user(kbuf, buf, count))
+		return -EFAULT;
+
+	kbuf[count] = '\0';
+
+	if (kbuf[0] == '1') {
+		timer_delete_sync(&blink_timer);
+		blink_active = 0;
+		gpiod_set_value(led_gpio, 1);
+		led_state = 1;
+	} else if (kbuf[0] == '0') {
+		timer_delete_sync(&blink_timer);
+		blink_active = 0;
+		gpiod_set_value(led_gpio, 0);
+		led_state = 0;
+	} else if (strncmp(kbuf, "b", 1) == 0 || strncmp(kbuf, "blink", 5) == 0) {
+		blink_active = 1;
+		mod_timer(&blink_timer, jiffies + msecs_to_jiffies(500));
+	}
+
+	return count;
+}
+
+static ssize_t gpio_led_read(struct file *file, char __user *buf, size_t count, loff_t *ppos)
+{
+	const char *state_str;
+	size_t len;
+	int state;
+
+	// Return 0 if the read offset is not zero
+	if (*ppos > 0)
+		return 0;
+
+	// Get actual GPIO value
+	state = gpiod_get_value(led_gpio);
+
+	if (state)
+		state_str = "on\n";
+	else
+		state_str = "off\n";
+
+	// Get message length
+	len = strlen(state_str) + 1;
+
+	// Copy to user space
+	if (copy_to_user(buf, state_str, len))
+		return -EFAULT;
+
+	*ppos = len; // update file offset
+
+	return len;
+}
+
+static struct file_operations fops = {
+	.owner = THIS_MODULE,
+	.write = gpio_led_write,
+	.read = gpio_led_read,
+};
+
+static int gpio_led_probe(struct platform_device *pdev)
+{
+	int ret;
+
+	led_gpio = gpiod_get(&pdev->dev, NULL, GPIOD_OUT_LOW);
+	if (IS_ERR(led_gpio)) {
+		dev_err(&pdev->dev, "Failed to get GPIO\n");
+		return PTR_ERR(led_gpio);
+	}
+
+	ret = alloc_chrdev_region(&dev_num, 0, 1, DEVICE_NAME);
+	if (ret)
+		goto err_free_gpio;
+
+	cdev_init(&my_cdev, &fops);
+	my_cdev.owner = THIS_MODULE;
+
+	ret = cdev_add(&my_cdev, dev_num, 1);
+	if (ret)
+		goto err_unregister_chrdev;
+
+	led_class = class_create("led_class");
+	if (IS_ERR(led_class)) {
+		ret = PTR_ERR(led_class);
+		goto err_del_cdev;
+	}
+
+	device_create(led_class, NULL, dev_num, NULL, DEVICE_NAME);
+
+	timer_setup(&blink_timer, blink_timer_callback, 0);
+
+	dev_info(&pdev->dev, "LED driver loaded\n");
+
+	return 0;
+
+err_del_cdev:
+	cdev_del(&my_cdev);
+err_unregister_chrdev:
+	unregister_chrdev_region(dev_num, 1);
+err_free_gpio:
+	gpiod_put(led_gpio);
+	return ret;
+}
+
+static void gpio_led_remove(struct platform_device *pdev)
+{
+	device_destroy(led_class, dev_num);
+	class_destroy(led_class);
+	cdev_del(&my_cdev);
+	unregister_chrdev_region(dev_num, 1);
+	gpiod_put(led_gpio);
+	timer_delete_sync(&blink_timer);
+}
+
+static const struct of_device_id gpio_led_dt_ids[] = {
+	{ .compatible = "lkss,gpio_led", },
+	{ /* sentinel */ },
+};
+
+static struct platform_driver gpio_led_driver = {
+	.probe = gpio_led_probe,
+	.remove = gpio_led_remove,
+	.driver = {
+		.name = "gpio_led",
+		.of_match_table = gpio_led_dt_ids,
+	},
+};
+
+MODULE_DEVICE_TABLE(of, gpio_led_dt_ids);
+
+module_platform_driver(gpio_led_driver);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Lab2 Ex3");
```

Comment:
keep consistent

Reply:
will update it - was a copy paste from ex3.
Actually ex4 and ex5 are extensions of ex3 - I've stated from that one and added extra "stuff" to it.